Demotion Consoles: lets heads demote crew out of their department

Kyep opened the pull request:
Adds Department Management Consoles to all head offices.

Features:
- Heads can reassign crew already in their department to other positions within their department (e.g: Engineer to Atmos Tech, or Sec Officer to Detective, or MD to Geneticist, etc)
- Heads can also reassign crew in their department to 'Civilian', but this is one-way. Their ID is updated to have normal 'Civilian' rank. Only HoP or Captain can re-hire the person after this - the other heads cannot. As such, this console only works for demotions and management within a department. Cross-department transfers still have to go through the HoP.
- Heads can also 'Demote' (new, red option) crew in their department. This acts like an assignment to Civilian, except that it sets their title to 'Unassigned', and strips any department-specific ID card skin they have. This is to make it clear to the HoP, should they ask to be hired somewhere else later, that they were fired by their previous head of staff. Note that Demotion is not the same as Termination. Termination removes everything, even their basic civilian/maint access. Demotion doesn't.

Heads cannot use these consoles to change job slots, affect any other department, fire themselves, etc. The consoles are quite limited in what they can do. **As of latest update, they cannot be used to hire anyone, either. They are strictly demotion consoles.**

Map:
![](http://i.imgur.com/EaN8H2j.png)

Updated HoS office image (after brig remap):
![](https://farm5.staticflickr.com/4202/34029102283_b143268404_b.jpg)

🆑 Kyep
add: Every head of staff now has a Department Management Console in their office. This is a very limited version of the ID console that they can use to demote people out of their departments, or move people between jobs in their department. These consoles cannot be used to hire anyone, or to accept cross-department transfers.
/🆑

tylerh95 commented:
@Kyep Nice, I like it. However, I feel that this kinda detracts from the HoP's job. Perhaps this could be something only located in the HoP office or simply a feature that is added to the identification computer?

Kyep commented:
@tylerh95 The entire point of this is to make it easier for heads to get problematic individuals removed from their departments. Making it something only the HoP has access to, defeats the entire point of adding it. I did consider altering the ID computer to have this functionality, but the ID computer is a big, complicated device. It was easier to create a small, dedicated device for this specific function.

FalseIncarnate commented:
Would it be possible to make the demotion terminal auto-generate an employment record for that shift so that when the demoted party inevitably attempts to con the HoP into re-instating their previous position's access, the HoP can see they were given the boot by their boss?

Kyep commented:
@FalseIncarnate So...
- Do HoPs read employment records?
- Before transferring a fired person back into the department, they *should* ask that head of staff. Who will say no.
- The head who fired them really should announce this firing over command radio, at least.
- I did think to have the terminal automatically broadcast a command radio message when it is used to demote someone. I took that out before putting up the PR, though.
- I also considered it converting the IDs demoted by it into old-style "data" icon IDs. Functionally the same, but with a different sprite that would tip off anyone looking at the ID that it wasn't normal.
- Lastly, I considered forcing demotions made via this console to free a job slot, and/or update the person's matching PDA... so it updates the manifest. If you have the ID, though, you probably also have the PDA, and you can always put the ID in and hit 'update' to update the manifest.

FlattestGuitar commented:
👍 to this. HoP is a horribly busy job with 120+ players online and there's barely any time in a round where you don't have a line of people waiting for you to assign them to that sweet janitor slot.

This will help unclog the line.

Would probably use custom sprites, though.

Kyep commented:
@FlattestGuitar If I had different sprites to use, I would. I'm not a spriter, though. Right now I'm just seeing if people like the concept. So far, everyone I've mentioned it to has.

Rumiluntti commented:
Why civilian? Why not terminated?
Would make more sense and actually discourages shitlers from shitlering.
You ARE fired after all

tylerh95 commented:
@Rumiluntti You're not fired, you're just demoted afaik

Kyep commented:
@Rumiluntti It is a demotion terminal, not a 'totally fired from NT employment' terminal. It is a 'get this guy out of my department' terminal, not a 'revoke ALL their access, even the maint access normally given to civilians!' terminal. It is a 'stuff heads can do by themselves' terminal, not a 'total replacement for HoP' terminal.

Twinmold93 commented:
With how much work the HoP normally has, and generally ignores the heads as is anyway, it'd be nice to have them be able to properly demote people out of their department instead of having a guy running around with access to a department they were fired from because the Head of Personnel didn't remove their access.

Lady-Luck commented:
I would ask you reconsider the immunities. Brig Physician is NOT joint medical, they just have medical access. And secondly, Geneticists being split is already such a gigantic issue for those who get hulk and misbehave that I would argue both heads need to be able to demote.

Birdtalon commented:
A good idea @Kyep HoP usually has a line which is too big to deal with so you either have to push to the front or wait for an excessive amount of time. 👍

Fox-McCloud commented:
As was pointed out a long time ago, with regards to this; departmental ID machines are a vastly superior option to this, as they can allow heads of staff to properly assign positions to individuals they need while also being able to demote them.

That's way better than this PR and would be better as it utilized pre-existing code.

Kyep commented:
@Lady-Luck Done.

Kyep commented:
https://github.com/tgstation/tgstation/pull/9940 https://github.com/tgstation/tgstation/pull/18297/files

uraniummeltdown commented:
I also think porting over tg's Department Management consoles would be better, a demotion only computer is kind of dumb to have, a management console is ultimately way more useful as you can demote people and add department access too.

pinatacolada commented:
Adding a note to the employment records about the demotion would be a good reason for the HoP to start reading them, even if just at glance

scrubmcnoob commented:
We definitely need this.

FalseIncarnate commented:
Prior to #6583 the HoP didn't have particularly easy access to employment records, so it was understandable they wouldn't have been checking them. Now, they have that easy access to those records, so checking them should hopefully become more common in some cases (though most people still probably haven't set theirs through character creation).

If this system were to auto-log demotions at least for the shift (editing their character creation preferences mid-round for multiple rounds gets into weird territory with persistence and the ability to just change it anyways), it would then at least encourage the HoP to check anyways just in case.

Necaladun commented:
Absolutely no reasons I can see not to have this. The only question is really about going further with the departmental access etc. :+1:

Lady-Luck commented:
I would rather have the Departmental Management as well, but I'm not gonna complain if we just get the demotion terminals.

Kyep commented:
@Lady-Luck @Fox-McCloud PR altered. 
# This is now a port of the Departmentmental Management Console. Heads can use this to hire Civs, and demote people from their dept. That's all they can use it for. Every head gets one in their office.

FreeStylaLT commented:
Should probably edit the title to reflect that

Fox-McCloud commented:
I don't think this is a port---calling it that without explicitly noting the differences between TG's implementation and our own is a bit deceiving.

Either case, big step in the right direction, but I still think you should go with TG's implementation here; it makes more sense and allows the heads the full control they need over assigning people to their individual departments.

SamHPurp commented:
With the change from demotion console to the ability to add civilians to the department, bypassing the HoP... I will be changing my position to not supporting this any more.

This removes a lot of their role, and more importantly removes a lot of oversight.

Suddenly all cultists have Science access, enabling their shenanigans even further, and other obvious use cases for why lack of oversight here is a bad idea.

orelbon commented:
I agree with sam if it was only a demotion console i'd be fine with it, but this is basically making the HOP redundant.

Aurorablade commented:
Yeah the civ.thing should stay with the hop and should be tgs implentation of.it

Crazylemon64 commented:
https://github.com/tgstation/tgstation/pull/9940
The source PR, for reference

Fox-McCloud commented:
>Suddenly all cultists have Science access, enabling their shenanigans even further, and other obvious use cases for why lack of oversight here is a bad idea

Which is really meaningless when you realize that all departments have guest passes anyway, which are even easier to use than this, can be spam generated, and last for such a long duration that they can be spammed without worry.

Kyep commented:
@Fox-McCloud 
The differences are:
- TG's console has horrible UI, ours doesn't. Really, really horrible TG UI.
- TG's console makes sounds upon every action, ours doesn't.
- TG's console doesn't let you assign actual jobs to people, only edit access. Ours is the reverse - you can assign jobs, but not individual accesses.
- TG's console lets you edit IDs of your level (e.g: one CE on another). Ours doesn't.
- TG's console supports the Job Priority System. Ours currently doesn't. I actually included that in the port, but Neca asked me to take it out, and make it a separate PR, simply so they could be discussed separately. So, I took it out before committing.

I did copy TG's card console implementation over here..... but I found the result was terrible. The UI was TG's terrible UI, it made sound every few seconds, it had a bunch of things I considered to be bugs, and it used a ton of procs that don't even exist in our codebase.
Given that I had to essentially rewrite the thing, basing it on our code was better. This let me keep the good Paradise UI, and was generally much smoother.

So, no, **I don't think that simply copying TG's implementation is a good idea.** If you want Department Management Consoles, you need to consider this version of them. The functionality is all there. At least as good as TG, if not better. Without reverting our UI to the 1990s style that TG uses.

SamHPurp commented:
@Fox-McCloud And they are totally arse to use, and only last 30 minutes max..

Aurorablade commented:
No it needs to leave civillien hireing into a department to the hop. I am all for making the ui better but this is taking away some core stuff from the hop I can only agree with this if it leaves putting a civi in the department with the hop. I MIGHT be okay with letting someone change just jobs WITHIN a department if no custom title is involved, otherwise it needs to be access only

Aurorablade commented:
Oh and edit meta map

Spartan6 commented:
So now the HoP is essentially useless and should be renamed to First Mate or Captain's Secretary. Sweet.

I preferred the demotion stuff, having this all managed by department heads is really neglecting the HoP now. Like, he now is only needed for Supply and Service. The two departments people don't like asking for.

Big nopey unless HoP gets something to do or is competently replaced with something.

Lady-Luck commented:
Yeah, no. If this change passes I'll actually be able to play HoP again without losing my mind.

SigholtStarsong commented:
Let's go ahead and remove HoP while we're at it - this PR effectively removes the need for the role altogether.

Lady-Luck commented:
That's absurd. The HoP is so glued to his ID computer that he can't even fulfill the chain of command properly if he needs to be captain and generally has to let the HoS do it. The HoP never has time to supervise or run the two departments under his direct purview.

This is not going to remove the need for the role even slightly.

SigholtStarsong commented:
What role does the HoP fill that IAA will not already? Servicon, as a
department, is already so insulated and compartmentalized that the HoP has
little-to-no need to legitimately oversee them, and Cargonia has the QM who
fills the role of "departmentmental" head.

Of you intend the HoP to function as HR, then he has to have the function
he currently fulfills. The computers here effectively make it so that the
HoP no longer needs to serve in an HR role, creating a situation where we
have a job that has no RP or mechanical reason to exist except as 2IC,
which, as you helpfully pointed out, is more-or-less going to end up being
the HoS.

On Mar 17, 2017 12:02 AM, "Lady-Luck" <notifications@github.com> wrote:

> That's absurd. The HoP is so glued to his ID computer that he can't even
> fulfill the chain of command properly if he needs to be captain and
> generally has to let the HoS do it. The HoP never has time to supervise or
> run the two departments under his direct purview.
>
> This is not going to remove the need for the role even slightly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ParadiseSS13/Paradise/pull/6766#issuecomment-287267591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APuF6voxqot5TrulX215ayp7gLd3XNRVks5rmhPjgaJpZM4McMnp>
> .
>

Lady-Luck commented:
What is 2IC? I'm assuming it's referring to the chain of command.

Have you played the server lately? HoPs are having to deal with Disneyland lines every single shift. The identification computer has turned into their only function. They don't have time for anything else. It's not meaningful, it's miserable.

Maybe service as a department is so insulated because the HoP never has time to run it, have you thought of that? And the Quartermaster isn't a Head of Staff, HoP still needs to provide oversight to make sure everything is running smoothly there.

They will have more time for extending security's access when need be. They will have more time to seek out filling necessary roles in the service department, such as more competent janitors. They will have more time to place important orders that will allow them to extend job slots without running into supply issues. They will have time to fulfill the chain of command.

There is so much the HoP is capable of that the HoP used to have the time to do, but high-pop has made the role into a computer monkey.

Lady-Luck commented:
Oh, and by the way, if I'm reading the description right, the HoP will still be instrumental in doing job changes from department to department. A doctor could technically get the CMO to demote them to civilian and then go to the RD and get promoted to scientist, but that is a lot more hassle than just going to the HoP desk.

The HoP will still have SOP requiring him to go to their desk when called, the CMO and RD will have no such obligation with their new ID computers.

SigholtStarsong commented:
What useful role does the HoP have in Servicon, even? "Batender, make more
drinks! Chef, cook food! Botany, stop growing nothing but weed and make
stuff for the chef!"

Servicon is kind of self-running by its very nature.

I've never seen an HoP ever give more access to Security - usually, it's
the Captain doing that. Same with the termination of executee's IDs.

The job slots argument doesn't hold much water, since a halfway competent
HoP can open more slots AND run the line as fast as he pleases.

Also, as far as I recall from SOP, the HoP doesn't take command if there's
a Stationwide emergency - it's supposed to be the HoS. The HoP only ever
takes over when the Captain SSDs.

On to your poimt about the QM - Why even have that role, then, if the HoP
is supposed to run that department as well as be an HR "computer monkey."

Semi-related, if the HoP isn't handing out paperwork and is, instead, just
hiring blindly,they're a shitty HoP.

On Mar 17, 2017 12:18 AM, "Lady-Luck" <notifications@github.com> wrote:

What is 2IC? I'm assuming it's referring to the chain of command.

Have you played the server lately? HoPs are having to deal with Disneyland
lines every single shift. The identification computer has turned into their
only function. They don't have time for anything else. It's not meaningful,
it's miserable.

Maybe service as a department is so insulated because the HoP never has
time to run it, have you thought of that? And the Quartermaster isn't a
Head of Staff, HoP still needs to provide oversight to make sure everything
is running smoothly there.

They will have more time for extending security's access when need be. They
will have more time to seek out filling necessary roles in the service
department, such as more competent janitors. They will have more time to
place important orders that will allow them to extend job slots without
running into supply issues. They will have time to fulfill the chain of
command.

There is so much the HoP is capable of that the HoP used to have the time
to do, but high-pop has made the role into a computer monkey.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/ParadiseSS13/Paradise/pull/6766#issuecomment-287269358>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/APuF6lfQ2zxsAyVpJo5YRmhR9wSknO5Qks5rmheugaJpZM4McMnp>
.

Lady-Luck commented:
>I've never seen an HoP ever give more access to Security - usually, it's the Captain doing that. Same with the termination of executee's IDs.

Why do you think that is? The HoP never has time. They are constantly servicing a line of people, endlessly. There used to be so much more to the job than that, but high-pop has stripped any other thing they may have wanted to do away.

>The job slots argument doesn't hold much water, since a halfway competent HoP can open more slots AND run the line as fast as he pleases.

It holds plenty of water, because you can't just increase some job slots all willy-nilly, if you don't have the supplies to actually make sure those new job slots get what they need, then it's actually a detriment to the department. You're best equipped to make the proper orders.

>Semi-related, if the HoP isn't handing out paperwork and is, instead, just hiring blindly,they're a shitty HoP.

Paperwork is not reasonable with the lines HoP gets now. Again, have you played the server at all recently? I've seen lines go around the corner at the start of the round more often than not, while people who want to be in line pass by and go 'fuck that, that line is too long'.

>On to your poimt about the QM - Why even have that role, then, if the HoP is supposed to run that department as well as be an HR "computer monkey."

The HoP has the final say in cargo, period. That's the way our SoP is written. They're not able to provide proper oversight to make sure the Quartermaster is doing their duty. I'm not going to argue with you about the command structure that's already in place in this PR, because it's not the place for it.

Twinmold93 commented:
@Kyep @Fox-McCloud 
There is no reason to allow a head to be able to hire a civilian/assistant (or anyone, really) who doesn't already have at least one of the accesses the head has for their department into their department for access adjustment or termination. Doing so removes the actual main job of the Head of Personnel. A head should only be able to adjust the access for the employees within their department (defined by the access being within their ID, with few exceptions) and fire them (which would remove all relative access to the department from the ID).

For example, the CMO cannot fire the Brig Physician (it's an exception), but can strip the Geneticist of Medical access (forcing the geneticist to only have science access plus access to their lab, of course). The CMO can fully demote a medical doctor of all medical related access, but if that doctor also has, say, bartender access, he keeps that access. The CMO would also be able to give chemistry access to another doctor, if chemistry is without staff. Lilly Sue Civilian walks in asking for a job, still has to go through the Head of Personnel or Captain.

Lady-Luck commented:
@Twinmold93 If these were the unlisted times I would agree with you. These are not the unlisted times. The workload of hiring so many civilians has become so incredibly immense that the duty should really be spread around to provide some relief to the job.

I used to love playing HoP. Now it's a mind-numbing experience because the line is fucking endless.

Lady-Luck commented:
I would again like to ask who here has played HoP when the server is at roughly 130 players or more at primetime?

Is sitting in the same place with the ID computer until the shuttle is called really enjoyable to you? The line only clearing when there's a stationwide disaster that pretty much makes all the work you put in completely moot?

HoP used to be one of my favorites. I would get dressed up in the HR outfit, make some templates for paperwork, interact with the people in line... now the workload is so great there is this massive pressure to do things as quickly and efficiently as possible, endlessly, and never really do anything else.

Twinmold93 commented:
@Lady-Luck And that line UTTERLY disappears if you port this PR as is. It's a huge lose of the environment of the game itself, and the point of the role. Suddenly, the Head of Personnel can just go 'line is never open, go to the heads to get a job'. And we lose any practical use for the position itself other than a free 'become Captain' card, since as noted, service and cargo are pretty self-running as is. Every time I've ever played Head of Personnel, I've never had to manage my actual departments. There isn't a need or point to. And it's nice for these departments because it gives people a chance to ease into the game without immediate fear of reprimand for a slight screw-up because your boss is actually watching.

Making it so the inner-department adjustments and demotions are done internally will help the mid and late-round situations that can happen, leaving the Head of Personnel with a little more time to do what they want or need. In the early-round, it's just that line. And if there are that many people in the line who are civilians looking for jobs, that actually brings to question the number of jobs we have available at the start of the shift in general, rather than the capability of the Head of Personnel, since it means we don't have enough low-end jobs available. The PRs to rectify this would greatly help with this in general, since you'd theoretically have less people getting put to civilian or HAVING to go civilian.

Lady-Luck commented:
>And that line UTTERLY disappears if you port this PR as is. It's a huge lose of the environment of the game itself, and the point of the role. Suddenly, the Head of Personnel can just go 'line is never open, go to the heads to get a job'.

7. The Head of Personnel may not leave their office unmanned if there are personnel waiting in line. Failure to respond to personnel with a legitimate request within ten (10) minutes, either via radio or in person, is to be considered a breach of Standard Operating Procedure;

Twinmold93 commented:
You're presuming that Standard Operating Procedure will even be adhered to for that condition, which very few people would forcibly follow (I know you are one of the few who would). However, SOP is a guideline for most players, not the way to play. And, if we support heads being able to hire on demand, this particular SOP has less hold than it does now.

Lady-Luck commented:
And please read the entire PR. This does not allow Heads of Staff to do department to department transfers, they can only hire those at civilian rank.

The Heads of Staff are not obligated to provide this service, whereas the HoP is. That's major.

The HoP is the only one that can hire to Service, and now perhaps Security will not feel so obligated to go straight to the Captain when they need their access extended for an investigation.

Lady-Luck commented:
>And, if we support heads being able to hire on demand, this particular SOP has less hold than it does now.

A department head has zero obligation to actually use these identification computers, whereas the HoP will still be obligated. Heads of departments are often very busy.

I ask again, have you played HoP both before we were listed and now that we are at 130 players? It's a world of difference. I don't think people arguing against this understand the crushing workload that playing a loyal, competent HoP is now.

Twinmold93 commented:
Two heads working together can easily do department to department transfers. Demote, then hire. I know there are a number of people who would take the shortest and quickest route just to get into a job, and any other method is useless to them. The quickest method would be walk to Head of Staff (or go to HoP, whom should be doing some sort of paperwork), find out they have to do paperwork, then go directly to the head of staff they were trying to be hired under and, depending on who is playing the Head (and let's face it, we have a lot of new players and people who just do the fastest action in heads right now) who would just hire on the spot, making the need to go to the HoP less than necessary and extra 'busy' work so to speak.

Look at both extremes of the situation, it's not one sided. Is there a lot of work right now, yes. I've seen the lines. There are, however, other ways to go about this than just an ID console, and yet it doesn't seem to have been acknowledged. More jobs in general assists greatly, because if there are 15 people in the line, and suddenly 10 more job slots available from round-start across the station, you can expect (with the people avoiding the line because line now being tempted to join it), to drop to about 7 to 9. Doesn't sound like much, until you take a consideration of actual percentages. And that's not including other positions.

At 130+ people, most departments are already full, for what they support. Adding even more people can make things very much crowded and make items less common and in-fighting worse. So, most of the actual available jobs are Service and Supply anyway (to fit comfortably, and even those positions are rather weak for additional personnel right now). You're almost guaranteed for 1 to 5 people to want to be (gateway) explorers, so the 7 to 9 becomes anywhere for 5 to 8 reasonably quick (again, adjusted because more people would get into a shorter line), which fall directly to the Head of Personnel and Captain's digression already.

Lady-Luck commented:
>Two heads working together can easily do department to department transfers. Demote, then hire. I know there are a number of people who would take the shortest and quickest route just to get into a job, and any other method is useless to them. The quickest method would be walk to Head of Staff (or go to HoP, whom should be doing some sort of paperwork), find out they have to do paperwork, then go directly to the head of staff they were trying to be hired under and, depending on who is playing the Head (and let's face it, we have a lot of new players and people who just do the fastest action in heads right now) who would just hire on the spot, making the need to go to the HoP less than necessary and extra 'busy' work so to speak.

If you're making the assumption that getting two heads to work in tandem would be easier than just going to the HoP then I don't know what to tell you. Especially seeing as if this change goes through, lines will definitely be shorter for the HoP and you wouldn't have to wait nearly as long.

I'll ask a third time. Have you played HoP both before we were listed and now that we are at 130 players?

I guarantee you, that if this change goes through, the HoP will still not be anywhere close to being bereft of important shit to do. The fact that hiring civilians has become the SOLE FUNCTION of the HoP is a PROBLEM, not the way things should be.

Again, there are so many important things a HoP could be doing. The fact that this is being viewed as making the job obsolete only serves to prove my point that the job has turned into a one-dimensional computer monkey.

HoP will be different after this change goes through, but it will be for the better.

Twinmold93 commented:
Making the assumption it's an impossibility doesn't bode well, either. Of course, you could just tell your boss you quit and they can fuck off and you're insta-fired. Go to another head, and if they AREN'T communicating properly, easy to get hired. Who said they had to actively be working together?

I played Head of Personnel before the list, and even a while into the listing. I haven't recently for personal reasons that have made me not really want to play games in general, but even I as an admin see rounds, floating around. Talking with dark, he's recently played, and has even state that it's the first 20 to 30 minutes that are hectic, then it starts getting really quiet.

With this extra information, it sounds like a forced extreme situation being presented to support the PR going through as is rather than a true practicality need. When I've been Captain recently multiple times, I've MANY times seen my Head of Personnel on the bridge, as the line hasn't been that bad that he had to dedicate any more than 25% of his time doing ID changes.

An extreme is what the argument is based on, and you still have yet to acknowledge the other ways to go about it.

Lady-Luck commented:
The first 30 minutes is when the other heads are going to be extremely busy too. You think they have time to play HoP when they're getting their department set up?

>Making the assumption it's an impossibility doesn't bode well, either. Of course, you could just tell your boss you quit and they can fuck off and you're insta-fired. Go to another head, and if they AREN'T communicating properly, easy to get hired. Who said they had to actively be working together?

Yeah uh, I would be extremely wary of someone dressed up in doctor gear at civilian rank suddenly wanting to join my department. I would probably not hire them based on that.

I will say I find your argument that this is going to make the HoP obsolete very improbable... and in the event it does happen, all it would take is another PR to correct the issue. Why are we so afraid to try things that could significantly improve gameplay in the days of auto-patching?

Twinmold93 commented:
> The first 30 minutes is when the other heads are going to be extremely busy too. You think they have time to play HoP when they're getting their department set up?
So, the same instance of if the head is too busy, they go to the Head of Personnel... That doesn't actually assist with their line, does it? If that was the intent in the original argument. A smart head would have that person turn in their gear immediately, and anyone with any idea to go to the locker room and change out has just made themselves appear as a civilian. Yes, for people who have played for a long time and know what to look for, it'd be obvious if they didn't change their outfit, but you also have to remember to think on much the same level as those without the experience.

> I will say I find your argument that this is going to make the HoP obsolete very improbable... and in the event it does happen, all it would take is another PR to correct the issue. Why are we so afraid to try things that could significantly improve gameplay in the days of auto-patching?
I would point out that literally every PR that has been considered since auto-patching could be set up and reverted with ease, yet still denies a good number of practical PRs. Using that as a reason of 'why not', if used here, should be usable and viable in all PRs. And it taken as is and that reason being even remotely probable for usage means ALL PRs should follow the same code of conduct.

Lady-Luck commented:
>So, the same instance of if the head is too busy, they go to the Head of Personnel... That doesn't actually assist with their line, does it? If that was the intent in the original argument.

You're thinking completely in binary here. There's not going to be a mass exodus of the HoP line, nor is nothing going to change.

>A smart head would have that person turn in their gear immediately, and anyone with any idea to go to the locker room and change out has just made themselves appear as a civilian. Yes, for people who have played for a long time and know what to look for, it'd be obvious if they didn't change their outfit, but you also have to remember to think on much the same level as those without the experience.

This is so much more of a hassle than just going to the HoP line.

>I would point out that literally every PR that has been considered since auto-patching could be set up and reverted with ease, yet still denies a good number of practical PRs. Using that as a reason of 'why not', if used here, should be usable and viable in all PRs. And it taken as is and that reason being even remotely probable for usage means ALL PRs should follow the same code of conduct.

That's not the point. The point is your argument hinges on something improbable - that it's going to make the HoP obsolete. So rather than giving it a shot and just tweaking it if the improbable does happen, the feature is being thrown out altogether. That's silly.

Kyep commented:
For reference...

Even with the PR as-is, the HoP still has many things that only they (or Captain) can do:
- Assign custom card skins
- Change job slots
- Transfer people between departments without having to get demoted by one head, then hired by the other
- Modify PDA skins
- Assign custom access beyond that which is default with a job (big deal, some jobs routinely ask for extra access)
- Assign custom job titles
- Terminate employees entirely
- Everything to do with the accounts terminal
- Manage the civil department, and the supply/cargo/mining areas

Personally, I do not care whether this PR keeps the "heads can hire people directly into their department" aspect. Nor do I care whether we use demotion consoles, or this light version of the Department Management Console. Both of those, I will go with popular opinion on.
It would take less than 10 minutes for me to modify these DMCs such that they do not permit heads to hire people from civilian into their department - and this function is reserved for the HoP again. I don't mind doing that, if it is what people want.
Personally, I prefer letting each head do the hiring/firing for their own department, because I think that makes more sense. It also frees up the HoP. However, I don't mind either way.

Two things to be aware of, though:
- I am not willing to simply copy over TG's code for ID consoles as is, because it is terrible. Its UI is an utter nightmare, doing things like displaying a complete list of jobs with no color-coding or sorting, not using nano/templates, etc. Functionally, it is half-assed, with only the ability to change access, not actual jobs, so you end up with, say, a Civ hired as a Sec Officer retaining their Civ ID, title, etc but getting shot by the rest of security when they're picking up their new job's gear from a locker. TG's implementation of ID consoles is terrible, both in looks, and in functionality. I consider the DMCs in this PR to be vastly superior, and I'm strongly against copying TG's lower-quality code, because it both looks worse, and also is functionally worse.
- I do think we should give this PR a chance, either in its current form, or in a "tweaked to still require hiring to go through the HoP" form. Letting heads directly fire people from their department, I think, would be a big improvement, without significantly impacting the HoP. I think almost everyone can agree on that. We should be willing to try things like this, and see how they play out. That requires giving them a chance, and not getting bogged down in endless debate over them.

Lady-Luck commented:
The 'Port' label should probably be removed from the PR.

Twinmold93 commented:
> You're thinking completely in binary here. There's not going to be a mass exodus of the HoP line, nor is nothing going to change.

If nothing is going to change, and the Head of Personnel is still going to have just as much work, then it's unnecessary code. Talking with kyet, a very good idea has come up. If they are going to be allowed to hire and fire, when demoted the ID assignment be set to Demoted or Unemployed, so another head cannot hire them without first going through the Head of Personnel. That covers the backdoor of quitting and re-hiring to bypass the Head of Personnel (and their paperwork).

Lady-Luck commented:
>If nothing is going to change, and the Head of Personnel is still going to have just as much work, then it's unnecessary code.

Re-read my comment. I specifically said that's binary thinking.

>Talking with kyet, a very good idea has come up. If they are going to be allowed to hire and fire, when demoted the ID assignment be set to Demoted or Unemployed, so another head cannot hire them without first going through the Head of Personnel. That covers the backdoor of quitting and re-hiring to bypass the Head of Personnel (and their paperwork).

Sounds fine to me. I don't think the 'backdoor' would've been a problem to begin with, but if it satisfies you then it's a great compromise.

Twinmold93 commented:
Binary thinking... Give some examples of when it'll be practical to have the heads, after that first thirty minutes, be able to hire civilians as they please (especially at 120+).

Lady-Luck commented:
Is there really a point in arguing this anymore?

Kyep commented:
Changes:
- Anyone demoted by a head is now given civilian status, with the custom title "Unassigned". This prevents a second head from hiring them, without them going through the HoP. This also means the HoP can tell the difference between normal civilians, and people who have been demoted.
- Console now resets demoted people IDs to the default skin.
- Fixed a bug where alt job titles were not considered as valid, even though the default titles were.
- Fixed a bug where deconstructing a console, then reconstructing it, broke it. This no longer happens.

Doukhan commented:
The recent plague of nazi HoP's makes this a necessity.
If there's a que a mile long you'd hope they'd lay off the paperwork.

king5327 commented:
There's a lot of commentary, but if the port was done right, the standard ID consoles should also have this functionality if a non-ID console access head tries sticking their ID into it. Furthermore, if modified via multitool to the console board, it is possible to make a console that is not locked to a particular department the way the CMO's, RD's, HoS's and CE's are. This means that unless a signed paper from the HoP or Cap comes to research, scientists are more or less free to hand out department management console boards since they're nowhere near as powerful as the ID console.

Aurorablade commented:
I am still against this if it is allowing heads to hire. That needs to be with the hop alone

Kyep commented:
As I have stated before, my main goal here is to allow heads to directly fire the worst offenders out of their departments. I do not care whether this PR keeps the "heads can hire people directly into their department, unless they were fired by another head" feature. To me, that's just a bonus feature that reduces the huge HoP lines we get on highpop rounds.

@FreeStylaLT @Necaladun @TullyBurnalot 
Please tell me your verdict on this.

scrubmcnoob commented:
will we reach month two in the next seven days before this is dealt with 🤔 
🅰️ 💯 chances we will 😎

Fruerlund commented:
status on this?

TullyBurnalot commented:
As discussed with @Kyep , PR is approved pending:

- Removal of the ability for Heads to hire from outside their department. This should remain a purely HoP thing to do;
- Changing the way it handles "Unassigned" personnel. Specifically, having separate buttons for assigning someone to the rank of Civilian and actually firing someone

Kyep commented:
@TullyBurnalot @Crazylemon64 @Fox-McCloud @tigercat2000 

The following updates have been made to this PR:
- Fixed map merge conflict.
- Heads can no longer hire civilians. They were already unable to hire anyone outside of their department. 
- Heads setting someone's job to "Civilian" no longer updates their card skin, or changes their title to 'Unassigned'. This is the soft "I am resigning from my job, no hard feelings" option.
- Heads now have a "demote" button on their consoles. This acts like an assignment to Civilian, except that it sets their title to 'Unassigned', and strips any department-specific ID card skin they have. This is to make it clear to the HoP, should they ask to be hired somewhere else later, that they were fired by their previous head of staff. 
- The PR's title and description have been updated to match its current status. A screenshot of the console in the HoS' new office has been added, too.

Currently supported features in this PR:
- Heads can reassign crew already in their department to other positions within their department (e.g: engineer to atmos tech, or Sec Officer to Detective, or MD to Geneticist, etc)
- Heads can also reassign crew in their department to 'Civilian', but this is one-way. Their ID is updated to have normal 'Civilian' rank. Only HoP or Captain can re-hire the person after this - the head cannot.
- Heads can also 'Demote' (new, red option) crew in their department. This is one way, and works as outlined above. Note that Demotion is not the same as Termination. E.g: Demoted people still get civilian (ie: maint) access. Terminated people don't.

SamHPurp commented:
Is there a secHUD pip for unassigned? Or are they automatically getting the black questionmark?

Kyep commented:
They get the Civ icon.
Unassigned is treated as an alt title for Civ.

SamHPurp commented:
👍

tigercat2000 commented on code/game/machinery/computer/card.dm:
```diff
@@ -302,13 +328,22 @@ var/time_last_changed_position = 0
 			if(is_authenticated(usr) && modify)
 				var/t1 = href_list["assign_target"]
+				if(target_dept && modify.assignment == "Unassigned")
+					src.visible_message("<span class='notice'>[src]: Demoted individuals must see the HoP for a new job.</span>")
```
No src. please, it is implied.

tigercat2000 commented on code/game/machinery/computer/card.dm:
```diff
@@ -302,13 +328,22 @@ var/time_last_changed_position = 0
+					src.visible_message("<span class='notice'>[src]: Demoted individuals must see the HoP for a new job.</span>")
+					return 0
+				if(!job_in_department(job_master.GetJob(modify.rank), FALSE))
+					src.visible_message("<span class='notice'>[src]: Cross-department job transfers must be done by the HoP.</span>")
```
No src. please, it is implied.

tigercat2000 commented on code/game/machinery/computer/card.dm:
```diff
@@ -393,17 +429,36 @@ var/time_last_changed_position = 0
+		if("demote")
+			if(is_authenticated(usr))
+				if(modify.assignment == "Unassigned")
+					src.visible_message("<span class='notice'>[src]: Unassigned crew cannot be demoted any further. If further action is warranted, ask the Captain about Termination.</span>")
```
No src. please, it is implied.

tigercat2000 commented on code/game/machinery/computer/card.dm:
```diff
@@ -393,17 +429,36 @@ var/time_last_changed_position = 0
+					src.visible_message("<span class='notice'>[src]: Unassigned crew cannot be demoted any further. If further action is warranted, ask the Captain about Termination.</span>")
+					return 0
+				if(!job_in_department(job_master.GetJob(modify.rank), FALSE))
+					src.visible_message("<span class='notice'>[src]: Heads may only demote members of their own department.</span>")
```
No src. please, it is implied.

tigercat2000 commented on code/game/machinery/computer/card.dm:
```diff
@@ -442,4 +499,43 @@ var/time_last_changed_position = 0
+
+/obj/machinery/computer/card/minor
+	name = "department management console"
+	target_dept = 1
```
Please create & use defines for all of these different target_dept values. I would recommend something like `TARGET_DEPARTMENT_ALL` through `TARGET_DEPARTMENT_x`

taukausanake commented:
So just wondering in passing, does the title Assistant mean anything anymore? I know it's literally the same as Civilian that is being used but I feel like the Assistant title is being phased out? Just a feel and I don't like having feels

SamHPurp commented:
@taukausanake Civilian is the root role name, and Assistant is an alternate.

Kyep commented:
@tigercat2000 Changed
